File: python/src/algorithms/hmac_algorithm.py
import hashlib
import hmac

from .algorithm import Algorithm


class HMACAlgorithm(Algorithm):
    def __init__(self):
        super().__init__(
            {"HS256": hashlib.sha256, "HS384": hashlib.sha384, "HS512": hashlib.sha512}
        )

    def generate_signature(
        self, encoded_header: str, encoded_payload: str, algorithm: str, secret_key: str
    ) -> str:
        return hmac.new(
            secret_key.encode(),
            f"{encoded_header}.{encoded_payload}".encode(),
            self.supported_algorithms[algorithm],
        ).hexdigest()

    def verify_signature(
        self,
        encoded_header: str,
        encoded_payload: str,
        algorithm: str,
        signature: str,
        secret_key: str,
    ) -> bool:
        return hmac.compare_digest(
            signature,
            self.generate_signature(
                encoded_header, encoded_payload, algorithm, secret_key
            ),
        )

File: javascript/src/algorithms/hmac_algorithm.js
import * as crypto from "crypto";
import { Algorithm } from "./algorithm.js";

export const HMACAlgorithm = function () {};

HMACAlgorithm.prototype = new Algorithm(
  { HS256: "sha256", HS384: "sha384", HS512: "sha512" },
  HMACAlgorithm
);

HMACAlgorithm.prototype.generateSignature = (
  encodedHeader,
  encodedPayload,
  algorithm,
  secretKey
) =>
  crypto
    .createHmac(
      HMACAlgorithm.prototype.supportedAlgorithms[algorithm],
      secretKey
    )
    .update(`${encodedHeader}.${encodedPayload}`)
    .digest("hex");

HMACAlgorithm.prototype.verifySignature = (
  encodedHeader,
  encodedPayload,
  algorithm,
  signature,
  secretKey
) =>
  crypto.timingSafeEqual(
    Buffer.from(signature),
    Buffer.from(
      HMACAlgorithm.prototype.generateSignature(
        encodedHeader,
        encodedPayload,
        algorithm,
        secretKey
      )
    )
  );
